src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Primitives
+{
+    public struct InplaceStringBuilder
+    {
+        private int _length;
+        private int _offset;
+        private bool _writing;
+        private string _value;
+
+        public InplaceStringBuilder(int length) : this()
+        {
+            _length = length;
+        }
+
+        public void IncrementLength(string s)
+        {
+            IncrementLength(s.Length);
+        }
+
+        public void IncrementLength(char c)
+        {
+            IncrementLength(1);
+        }
+
+        public void IncrementLength(int length)
```

Comment:
It's different because you are required to `IncrementLength` for all strings before appending them and I felt like having a method is less error prone then incrementing property.